### ISSUE TITLE
go/ir: mistakenly converts slice index to int on 32 bit systems

### ISSUE THREAD

**Issue by dominikh** [dominikh/go-tools#1175]
Input: 
```
package main

import "math"

func main() {
	var idx int64 = math.MaxInt32*2 + 8
	x := make([]int, 16)
	println(x[idx])
	println(x[int(idx)])
}
```

IR:

```
b0: # entry
	t1 = Const <int64> {4294967302}
	t2 = Const <int> {16}
	t3 = HeapAlloc <*[16]int>
	t4 = Slice <[]int> t3 <nil> t2 <nil>
	t5 = Convert <int> t1
	t6 = IndexAddr <*int> t4 t5
	t8 = Load <int> t6
	t9 = Call <()> println t8
	t10 = Convert <int> t1
	t11 = IndexAddr <*int> t4 t10
	t13 = Load <int> t11
	t14 = Call <()> println t13
	Jump → b1

b1: ← b0 # exit
	Return
```

go/ir converts the int64 to int before using it in IndexAddr. But on 32-bit systems that's not correct. With `GOARCH=386 go run`, the first indexing operation results in `panic: runtime error: index out of range [4294967302] with length 16` while the second operation succeeds, accessing `x[6]`.

**Comment by dominikh** [comment]
This still happens when indexing non-addressable arrays.

**Comment by dominikh** [comment]
A similar bug exists with shifts with signed rhs, where the rhs gets converted to uint64.

**Comment by dominikh** [comment]
We fixed both of these when we updated go/ir for generics.

**Comment by dominikh** [comment]
Look into https://go-review.googlesource.com/c/tools/+/400394/, too

**Comment by dominikh** [comment]
> Look into https://go-review.googlesource.com/c/tools/+/400394/, too

This was fixed by d52656bf730e7d55dc3f838ac4ed86838a56a5de.
